oximeter/db/src/oxql/ast/table_ops/limit.rs

```diff
@@ -0,0 +1,263 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! An AST node apply limiting timeseries operations.
+
+// Copyright 2024 Oxide Computer Company
+
+use crate::oxql::point::Points;
+use crate::oxql::point::ValueArray;
+use crate::oxql::point::Values;
+use crate::oxql::Error;
+use crate::oxql::Table;
+use crate::oxql::Timeseries;
+use std::num::NonZeroUsize;
+
+/// The kind of limiting operation
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum LimitKind {
+    /// Limit the timeseries to the first points.
+    First,
+    /// Limit the timeseries to the last points.
+    Last,
+}
+
+/// A table operation limiting a timeseries to a number of points.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct Limit {
+    /// The kind of limit
+    pub kind: LimitKind,
+    /// The number of points the timeseries is limited to.
+    pub count: NonZeroUsize,
+}
+impl Limit {
+    /// Apply the limit operation to the input tables.
+    pub(crate) fn apply(&self, tables: &[Table]) -> Result<Vec<Table>, Error> {
+        if tables.is_empty() {
+            return Ok(vec![]);
+        }
+
+        tables
+            .iter()
+            .map(|table| {
+                let timeseries = table.iter().map(|timeseries| {
+                    let input_points = &timeseries.points;
+
+                    // Compute the slice indices for this timeseries.
+                    let (start, end) = match self.kind {
+                        LimitKind::First => {
+                            // The count in the limit operation should not be
+                            // larger than the number of data points.
+                            let end = input_points.len().min(self.count.get());
+                            (0, end)
+                        }
+                        LimitKind::Last => {
+                            // When taking the last k points, we need to
+                            // subtract the count from the end of the array,
+                            // taking care that we don't panic if the count is
+                            // larger than the number of data points.
+                            let start = input_points
+                                .len()
+                                .saturating_sub(self.count.get());
```

Comment:
A nitpick, but I had to go look a few things up to make sure I understood this, whereas this is self-explanatory and doesn't rely on the fact that it's a usize and therefore intrinsically 0-bounded.

```rs
let start = (end - self.count.get()).max(0)
```

Reply:
Oh. On second thought, the fact that it's a usize is relevant because if you subtract more from it it will die immediately. 🤦 never mind

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6f45cdc8e262c24d2377e39f10958ebb

Reply:
Yeah, the whole `saturating_sub()` business is to avoid any conditionals or panics if the count is greater than the length of the array. Nice tool for this kind of thing.